File: Algorithms/1239_Maximum_Length_of_a_Concatenated_String_with_Unique_Characters/Java/MaximumLengthofaConcatenatedStringwithUniqueCharactersSolution1.java

// Space: O(n)
// Time: O(2^n)

class Solution {
    private Map<Character, Integer> cache = new HashMap<>();

    public int maxLength(List<String> arr) {
        if (arr.size() == 1) return arr.get(0).length();
        for (int i = 97; i < 97 + 26; i++) {
            cache.put((char) i, 1);
        }
        return backTracking(arr,0,0,0,false);
    }

    private int backTracking(List<String> arr, int curIndex, int tempCount, int res, boolean flag) {
        res = Math.max(res, tempCount);

        for (int i = curIndex; i < arr.size(); i++) {
            for (Character c : arr.get(i).toCharArray()) {
                cache.put(c, cache.get(c) - 1);
                if (cache.get(c) < 0) flag = true;
            }
            if (!flag) {
                tempCount += arr.get(i).length();
                res = backTracking(arr, i + 1, tempCount, res, flag);
                tempCount -= arr.get(i).length();
            }
            for (Character c : arr.get(i).toCharArray()) {
                cache.put(c, cache.get(c) + 1);
            }
            flag = false;
        }
        return res;
    }
}

File: Algorithms/1239_Maximum_Length_of_a_Concatenated_String_with_Unique_Characters/Python/Maximum_Length_of_a_Concatenated_String_with_Unique_Characters_Solution_1.py

# Space: O(n)
# Time: O(2^n)


class Solution:
    def maxLength(self, arr):
        length = len(arr)
        if length == 1: return len(arr[0])
        self.cache = {chr(i): 1 for i in range(97, 97 + 26)}
        self.flag, self.temp_count, self.res = False, 0, 0

        def bfs(arr, temp_res, cur_index):
            self.res = max(self.res, self.temp_count)

            for i in range(cur_index, len(arr)):
                for c in arr[i]:
                    self.cache[c] -= 1
                    if self.cache[c] < 0: self.flag = True
                if not self.flag:
                    self.temp_count += len(arr[i])
                    bfs(arr, temp_res, i + 1)
                    self.temp_count -= len(arr[i])
                for c in arr[i]:
                    self.cache[c] += 1
                self.flag = False

        bfs(arr, 0, 0)
        return self.res





